journal/src/main/java/pl/wsztajerowski/journal/records/DoubleBatch.java

```diff
@@ -0,0 +1,152 @@
+        if (isClosed) {
+            throw new IllegalStateException("DoubleBatch is closed");
+        }
+        while (true) {
+            batchLock.lock();
+            try {
+                if (isClosed) {
+                    throw new IllegalStateException("DoubleBatch is closed");
+                }
+                Batch activeBatch = currentBatch;
+                if (!activeBatch.hasRemaining(buffer.remaining())) {
+                    trySwapBatch();
+                    continue;
+                }
+                long offset = activeBatch.write(buffer);
+                if (activeBatch.isFull()) {
+                    trySwapBatch();                                                 // signal(consumer) - producer still has lock
+                }
+                if (waitForFlush) {                                                 // async write
+                    while (!activeBatch.hasBatchFlushed()) {
+                        activeBatch.getBatchHasFlushedCondition().await();
+                    }
+                }
+                return offset;
+
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            } finally {
+                batchLock.unlock();
+            }
+        }
+    }
+
+    private void runBatchConsumer() {
+        while (!isClosed) {
+            batchLock.lock();
+            try {
+                Batch batchToFlush = currentBatch;
+                while (batchToFlush.isEmpty()) {
+                    batchIsFullCondition.await();
+                }
+
+                flushBatchAndSignalAllWaitingWriters(batchToFlush);
+            } catch (InterruptedException | IOException e) {
+                throw new RuntimeException(e);
+            } finally {
+                batchLock.unlock();
+            }
+        }
+        // flush another batch if there is something to write
+        batchLock.lock();
+        try {
+            flushBatchAndSignalAllWaitingWriters(currentBatch);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        } finally {
+            batchLock.unlock();
+        }
+    }
+
+    private void flushBatchAndSignalAllWaitingWriters(Batch batchToFlush) throws IOException {
+        ByteBuffer writableBuffer = batchToFlush.writableBuffer();
+        int expectedBytesToWrite = writableBuffer.limit();
+        int bytesWritten = writeChannel.write(writableBuffer);
+        if (bytesWritten != expectedBytesToWrite) {
+            throw new RuntimeException("Written bytes mismatch - expected: " + expectedBytesToWrite + ", actual: " + bytesWritten);
+        }
+        batchToFlush.clear();
+        batchToFlush.markBatchAsFlushed();
+        batchToFlush.getBatchHasFlushedCondition().signalAll();
+    }
+
+    private void trySwapBatch() {
+        batchIsFullCondition.signal();
+        Batch nextBatch = currentBatch == batchA ? batchB : batchA;
+        if (nextBatch.isEmpty()) {
```

Comment:
why not single method, on Batch called, resetFlushMark, which checks isEmpty and resets if needed and return true?

if(nextBatch.resetFlushMark()){
  currentBatch = nextBatch;
}